### PR TITLE
Reduce service worker functionality to minimal offline page

### PR DESCRIPTION
**Pull request by benhalpern**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR strips out some of our _very cool_ [ReadableStream](https://dev.to/devteam/instant-webpages-and-terabytes-of-data-savings-through-the-magic-of-service-workers-1mkc) service worker functionality because it has been causing too many bugs related to the deployment of new code.

The upside of this functionality is _very instant_ initial page loads, but the downside to the current implementation is too many instances of difficult-to-predict deployment cache issues. These were always painful to deal with, but are exceptionally difficult in a Forem world where our devops cannot act quickly to fix production issues the same way we can with a single deploy.

The fundamental issue with our current implementation is just that it is _not exhaustive_ in terms of dealing with certain scenarios, and dealing with edge cases creates a lot of complexity.

We suspect there are still UX benefits and SEO benefits to handling basic offline functionality...

Screen Shot 2021-02-26 at 4 47 54 PM

And we still have our service worker lifecycle on top of which we may be able to methodically add back certain functionality that this PR removes, but we can do so with much more of an emphasis on ensuring all possible scenarios are accounted for so that we do not allow users to get caught in buggy situations.

One additional adjustment I made was just removing the "DEV" portion of the image on the offline page (and some code cleanup there)... We may want to offer a new fun way for Forem admins to customize this page, but I figured this small adjustment to make for a more minimal and generalized offline page made sense here.

## Follow up

There is additional code that can be removed with this new approach, but due to the nature of how service workers run code already downloaded to user browsers (i.e. the cause of this complexity in the first place), it makes sense to wait at least a few weeks to remove some of the other code that is not needed anymore in case it could break old installs.

## Related Tickets & Documents

https://github.com/forem/internalEngineering/issues/333

## QA Instructions, Screenshots, Recordings

This is mostly an adoption of this functionality: https://web.dev/offline-fallback-page/

Eyeball the functionality and test different network conditions to ensure this does not cause new problems.

### UI accessibility concerns?

This should not cause new a11y issues.

## Added tests?

- [ ] Yes
- [x] No, and this is why: I'm not really sure how to test this.
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [x] I've updated the [Developer Docs]

## [optional] What gif best describes this PR or how it makes you feel?

goodbye

### CODE REVIEW

**Review by rhymes — COMMENTED** [review]
I wasn't able to test this locally, when I trigger the Offline mode in Chrome's network Panel, the service worker doesn't kick in and I get the offline dino. Am I missing something?

sw

File: app/views/service_worker/index.js.erb [hunk 1]
```diff
@@ -1,229 +1,73 @@
-// Serviceworkers file. This code gets installed in users browsers and runs code before the request is made.
+// Serviceworkers file. Renders as serviceworker.js
+
 <% unless Rails.env.test? || ENV["SKIP_SERVICEWORKERS"] == "true" %>
-  const staticCacheName = 'static-1.4';
-  const expectedCaches = [
-    staticCacheName
-  ];
+  const OFFLINE_URL = "offline.html";
```
[the remark this answers is not in the document]
Comment (benhalpern) [reply app/views/service_worker/index.js.erb:5]: Added it back 👍

**Comment by benhalpern** [comment]
@rhymes Here is mine working:

https://user-images.githubusercontent.com/3102842/109430403-46ec1400-79cf-11eb-9049-40c0843848c1.mov

Did you possibly have `ENV["SKIP_SERVICEWORKERS"] == "true"`? in your env? We still have that line in there in case people want to turn of SW in dev (though this will make local service workers less of a pain)

File: app/views/service_worker/index.js.erb [hunk 2]
```diff
@@ -1,229 +1,75 @@
...
-
-    }
+    // If our if() condition is false, then this fetch handler won't intercept the
+    // request. If there are any other fetch handlers registered, they will get a
+    // chance to call event.respondWith(). If no fetch handlers call
+    // event.respondWith(), the request will be handled by the browser as if there
+    // were no service worker involvement.
   });
-<% end %>
+<% end %>
```
Comment (citizen428) [inline app/views/service_worker/index.js.erb:75]: Nitpick: this is complaining about a missing final EOL.

**Review by citizen428 — COMMENTED** [review]
Tested locally, seems to work well for me.

**Comment by rhymes** [comment]
@benhalpern no, it's to false and I can see the compiled JS. It's weird. I tried both in Chrome and Edge. The code and the logic look good though :)

**Review by mstruve — COMMENTED** [review]
I believe we can also get rid of the `async_info/shell_version` endpoint now.

**Comment by benhalpern** [comment]
> I believe we can also get rid of the async_info/shell_version endpoint now.

@mstruve We'll still want the old stuff to stick around for a while to support anyone who still has the old service workers installed. I covered that in the `Follow up` section in the writeup. I didn't get specific by what "a few weeks" could look like, but I think it shouldn't cause problems to keep it around, and later this month we can head back in and remove all that legacy code.

**Review by atsmith813 — APPROVED** [review]

**Comment by mstruve** [comment]
🤦‍♀️  Thats what I get for not reading the full description. @benhalpern can we put a `TODO` comment in the code around what needs to be removed that way when someone comes across the code they are not confused as to why they dont see it used anywhere?

**Comment by benhalpern** [comment]
@mstruve I added a bunch of TODO comments in most of the places. Because of [this PR](https://github.com/forem/forem/pull/12851) I stayed clear of `routes` because it seems like it could create some annoying merge conflicts.

**Review by mstruve — APPROVED** [review]
LGTM! thanks for adding those TODOs, that will make our job in mid march easy peasy. You have a failing spec that I cant quite make heads or tails of.

**Pull request merged**